### DISCUSSION TITLE
New features! Clipboard paste and additional features for free text annotations.

### DISCUSSION THREAD

**Discussion by youmjb** [sumatrapdfreader/sumatrapdf#3489] [General]
- I created a feature to paste clipboard images into PDF documents and a feature to resize free text annotations to the size of the text.

The added features are as follows

![화면 캡처 2023-05-12 171438](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/fe9d3ad4-6cb8-4640-accd-0fd7130f7ee1)

![화면 캡처 2023-05-12 171529](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/1239068e-9273-47e9-81d3-b15d650d2664)

1.  Copy one image from your browser or other program, then select "Paste Clipboard" from the PDF document, as shown in the first embedded image above, and the image will be pasted into the PDF file, as shown in the second embedded image above.
2. When you create a free text annotation, writing a text is automatically available, and the text annotation is automatically resized according to the length of whole sentence in the text annotation. 

- To implement this feature, please download the files from below link(my git repository) and replace the source code files of the original SumatraPDF with these files, and compile them to implement the feature. Or, you can just download the 'SumatraPDF.exe' from my git repository and use it.

https://github.com/youmjb/SumatraPDF-mod.git

**Comment by DaaBoss** [comment]
Although it sounds cool, hopefully something like this will end up in the main product as a feature.  (I'd like it, but don't need it enough to take the initiative to use your changes.)

Visiting forums today to:  Suggest specific changes that are required to make annotations as a useful tool for most purposes.  T these apply to not just your changes, but the main code path as well, (so re-post these elsewhere, or I might do this later, myself):

1. --- Size matters -- I annotated a one page PDF with one small change.  Saving annotations does NOT balloon the file size, but then anyone can see the underlying object.  The only option to prevent discovery of what I covered up, was to PRINT AS PDF, but then the size went from 800k to 6,800k -- Almost 10x as large, just to add two words of text, "-- REDACTED TEXT --".  Not sure how to fix this, but often, this ballooning size prevents widespread use.  FYI, Adobe Reader (PRIOR to DCE), also balloons the size. with PRINT TO PDF.
2. --- Resizing any annotation BOX I place in a PDF is a requirement, in most cases.  What are the odds it will fit where I need it, even if it is just text??  In my case, I wanted to paste ALL the sales info into an INVOICE.  When I figured out I couldn't resize, I can't see how I'm going to present it so it isn't a formatting mess, without resize.  What if it isn't large enough?  What if I have room, but need to not overlay existing text?  (Resizing the TEXT box, or added graphic, not the text FONT).
3. --- VERY often, we need to take an existing doc and REDACT just a part of it, and need a tool to do so.  Example:  I have a doc that has a SSec # showing that I need to remove.  I need two things:  
 --- A --->  To OBSCURE the underlying text, (or REMOVE the TEXT from the FILE, assuming it was not a graphic of the SSec # ).  
 --- B --->  I then need to SAVE the doc so you cannot UNCOVER / DISCOVER the underlying SSec#. 
  I did these two things by adding four "text annotations" yellow "?" boxes, which covered the SSec # section, (the whole doc was a graphic), without need for any resizing.  But then to prevent discovery, I had to PRINT-TO-PDF, ballooning the size. 

So, these are my requirements for what I'd like to see for annotations.  
1. -- Easily obscuring part of a PDF, using a resizable area.  
2. -- Preventing that obscured part from recovery from the file itself if sent via email, etc..  
3. -- Don't balloon the size when saving and MUST ALSO remove the obscured part.
4. -- Easily ADD a RESIZABLE text box, (or picture from paste).
5. -- Added bonus -- a NEW TYPE of text box -- in the TEXT BOX BORDER at the top, (and bottom?),  replace "------" with "----- This section was added by: ________, dated today: ___/___/____, original date of doc was ___/___/____. ------".   Sometimes it is not apparent what was original, and what was added in the instance of annotating an invoice.  File-date is changed when saved, so record that manually when required.

**Comment by GitHubRulesOK** [comment]
@DaaBoss 

4) is what above alteration attempts

`1) 2) 3) 5)` Redaction requires a full blown editor to totally restructure a PDF since overstrike as you notice does not remove contents, in fact delete an annotation is very easy and then the file is back to the way it was before signing or any other overlaid changes.

Annotation in PDF is as Adobe design simply an appendage to core file  used for collaborative redlining.

Agreed overprint can work in limited cases but there are still ways to guess what was overprinted. Redaction is not the best way to alter a source where data should not have been included, See many redaction embarrassments in the public domain, Like recent EEC/Zenica fiasco.

**Comment by GitHubRulesOK** [comment]
@youmjb 
I have not tested but sounds like it closes many open issues so best provided as a "Pull Request"

**Reply by youmjb** [reply]
> @youmjb I have not tested but sounds like it closes many open issues so best provided as a "Pull Request"

I realized that the executable file is not available for download, even though I put it on github. As for "pull requesting" it, I'll get it up and running.

**Reply by GitHubRulesOK** [reply]
I downloaded the exe so you should note that placing custom images is problematic as I crashed by moving the image around the page NOTE the image shows as `Caret` in editor (should it not be `stamp` or similar widget)
Exception: C0000005 ACCESS_VIOLATION
Faulting IP: 00007FFA0C513526 01:0000000000062526 ntdll.dll!RtlWaitOnAddress+0xc6
Fault writing address 0000000000000024

Hmm second run and not now crashing ? but image disappears even though object is draggable around the page, so there is a rendering issue of some form.

**Reply by GitHubRulesOK** [reply]
The text box looks like a very good improvement I like the borderless default
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/8dbb1b47-fe92-4fd4-9396-6291347860ff)
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/fd5a2597-bfde-49f4-b63f-ab03acab189b)

**Reply by youmjb** [reply]
> The textbox looks like a very good improvement, I like the borderless preset ![image](https://user-images.githubusercontent.com/12754913/238103268-8dbb1b47-fe92-4fd4-9396-6291347860ff.png) ![image](https://user-images.githubusercontent.com/12754913/238103323-fd5a2597-bfde-49f4-b63f-ab03acab189b.png)

![스크린샷 2023-05-13 181706](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/519e5cc9-e7d4-40c3-816a-6fb9fd5a74bf)

Hi! I made some revisions.

- I improved the code to avoid crashing when moving the image object around the page. 
- The image clipboard is now resizable based on the width of the image. You can decrese or increase the image width using the trackbar on the annotation window.
- Some bugs related to the free text annotation were removed.

I also created a PDF file (Customization of SumatraPDF.pdf) showing what I changed from the original code and uploaded it to my git repository.

https://github.com/youmjb/SumatraPDF-mod.git

Thank you for your feedback!

**Reply by GitHubRulesOK** [reply]
Agreed its better but I still have issues with it disappearing, which may be due to the `caret` tag rather than a stamp identity, so if I take the file to a PDF editor the "blend" seems to be different to its placement of an image stamp. let me see if I can show the difference.

when you paste the image if you enter text in the comments box before save the image disappears

**Reply by youmjb** [reply]
> I agree that it is better, but I still have problems with the disappearance of the image, which is due to the "caret" tag rather than the identity of the stamp. So when I put the file into a PDF editor, the "blend" seems to be different than placing an image stamp. Let me see if I can show the difference.

![스크린샷 2023-05-13 195822](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/7b6d24dd-fc91-4a85-b195-bc10763233d2)

This is a screenshot from the last run I built. As you can see in the Annotation Window on the right, I removed all objects except for the image width so that the image object doesn't disappear. Please check it out. Thanks.

**Reply by GitHubRulesOK** [reply]
I am not the developer just a tester, so cant say what should or should not be done, but my observation for the image placements are the initial scale is roughly twice the expected size since placed as roughly pixels per point (not a big problem as you have introduced a scaling slider so that's good), the bigger issue I have is that conventionally the stamp can cary a text comment too, here is a sample with a stamp from xchange and 2 stamps from second exe, so those need different comments to distinguish which is the signature and which is the approved stamp.

[sample-hyperjump-stamped.pdf](https://github.com/sumatrapdfreader/sumatrapdf/files/11469507/sample-hyperjump-stamped.pdf)

**Reply by youmjb** [reply]
> I am not the developer just a tester, so cant say what should or should not be done, but my observation for the image placements are the initial scale is roughly twice the expected size since placed as roughly pixels per point (not a big problem as you have introduced a scaling slider so that's good), the bigger issue I have is that conventionally the stamp can cary a text comment too, here is a sample with a stamp from xchange and 2 stamps from second exe, so those need different comments to distinguish which is the signature and which is the approved stamp.
> 
> [sample-hyperjump-stamped.pdf](https://github.com/sumatrapdfreader/sumatrapdf/files/11469507/sample-hyperjump-stamped.pdf)

Your suggestion is very appropriate. I've been utilizing Caret because it's not as useful as I thought it would be, but either creating a whole new category (image category?) or utilizing the stamp would be a radical solution. I will try to develop a new way soon.

**Reply by GitHubRulesOK** [reply]
I suggest that a Pull request for your text box could be favourably considered whilst your separate improvement for images is in progress

**Reply by youmjb** [reply]
> I suggest that a Pull request for your text box could be favourably considered whilst your separate improvement for images is in progress

Now, the pasted image object can carry the text comment. I revised the code. The functionality of free text annotation is also improved. The text now fit to the container with all the border thickness and font size. 

https://github.com/youmjb/SumatraPDF-mod.git

**Reply by GitHubRulesOK** [reply]
Sorry for delay but had some other issues to attend to so It looks good so far, Definitely much improved over official pre-release. and addresses (I will say partially, but agree its majority of) requested improvements on existing annots.
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/1f9f50ba-1a00-4694-b322-fd37960c75f0)

Minor gripe it still says caret rather than stamp and as such it degrades to caret in other editors, but without looking closer at construct can not t say how that may be handled. Other issue with text is no easy way to tug shape (e.g. handles) but I had opened a feature request to be able to edit the dx dy values for that.

**Not your doing** but the way MuPDF generate Text Box leaves an issue in other editors, where text is defaulted to a different shape ! 
For Example if I add more text in MS Edge It will scale down font , "re-wrap" lines and "resize the shape" so its not easy to re-alter dx and dy when returned to SumatraPDF
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/75463b20-7977-4e2f-8538-a125308dce5b)

The Problem with the paste degenerating to caret when moded in say for example Tracker xchange is shown here so although not tagged as caret that's what it will be shown on touch up. whilst the stamped "Draft" keeps its Appearance.
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/f4f79087-7150-49fb-b94e-41db158ca9dd)

Thus on return to SumatraPDF its downscaled as modified to caret and unless say recolour as red will be hard to spot
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/7b14c134-365e-480a-a413-2791e4c7a2cf)

**Reply by youmjb** [reply]
![스크린샷 2023-05-16 064643](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/0f62f8eb-43ec-4b3f-aad0-6a748b155995)

![스크린샷 2023-05-16 091331](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/0cb5b98c-7640-459a-a417-ad6add4e35f1)


That's really good feedback. I appreciate it. 
I made a new category 'Image' object and brought back the 'Caret' feature. You can see the above screenshot which demonstrates both the 'Image' and 'Caret' features.

I also added a few improvements to the free text annotator. 
1. I fixed an issue where typing 'Enter' key doubles the line spacing. 
2. I fixed an issue where increasing the Borderline thickness would cut off the text. 
3. I Adjusted the placement of the text to be more centered.

https://github.com/youmjb/SumatraPDF-mod.git

**Reply by GitHubRulesOK** [reply]
Wow that seems to close out my observations as the image is now a distinct (Unique to SumatraPDF ?) addition 
not seen as a discrete movable annotation object in Tracker to be altered but then that's not a SumatraPDF issue ?.

Text looks good and again not your fault, but a new issue raised since you started has been raised in https://github.com/sumatrapdfreader/sumatrapdf/issues/3502, I suspect its part of the way annotation edits are handled and don't expect that you could absorb/close out  that issue

**Reply by youmjb** [reply]
KJK released a version with a new feature that allows moving tabs around, I applied my works to that version. 
I hope this helps.

https://github.com/youmjb/SumatraPDF-mod.git

**Reply by GitHubRulesOK** [reply]
@kjk comments ?

**Reply by youmjb** [reply]
> @kjk comments ?
![스크린샷 2023-05-27 134107](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/c07f5d1e-41c4-4445-9e2d-898b64e42975)

![스크린샷 2023-05-27 134230](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/5372d99b-fa5c-4132-a76d-1c77245259c5)

I have added a new feature and improved stability to prevent crashes.
1. As shown in the screenshot above, I have implemented the functionality to select a text area and draw a Text Box by right-clicking with the mouse.
2. Double-click the left mouse button to put text to the PDF, and the text will be displayed on the PDF page in real time as one is typing.

https://github.com/youmjb/SumatraPDF-mod

**Reply by GitHubRulesOK** [reply]
@youmjb 
I have problems with this new version since on open a file and left click it instantly triggers place text box, Now ***that may be because I have another app monitoring selections in SumatraPDF*** active, but its a point to be aware of that left click default may not always be as expected.
The default for Left click is altered in certain cases, so for LaTeX users double left click triggers Synctex support but without Synctex it attempts to select the current string between non alpha characters. In my augmented case it looks for a longer selection string via ahk, thus my AHK on first click "arms" a monitor for second click.

Your new version double  click ***without my enhancement*** active
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/1652a37a-92a2-418a-b757-159596e4b567)

Current pre-release double click **with a latex file** should go to this line in notepad
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/ec2008be-5046-4b85-94b1-002985120345)

I attach a demo set so you can see **normal pre-release behaviour** unpack this ZIP and drop sync.pdf onto the pre-release exe then double click a word should open notepad with the source line. (It may be **@start of non highlighted notepad line** as I have not yet decided what my latex editor should do.) In a real LaTeX IDE the behaviour can be variable.

[LATeX and Reverse-Search-textpad-demo.zip](https://github.com/sumatrapdfreader/sumatrapdf/files/11581959/LATeX.and.Reverse-Search-textpad-demo.zip)

**Reply by youmjb** [reply]
> @youmjb I have problems with this new version since on open a file and left click it instantly triggers place text box, Now _**that may be because I have another app monitoring selections in SumatraPDF**_ active, but its a point to be aware of that left click default may not always be as expected. The default for Left click is altered in certain cases, so for LaTeX users double left click triggers Synctex support but without Synctex it attempts to select the current string between non alpha characters. In my augmented case it looks for a longer selection string via ahk, thus my AHK on first click "arms" a monitor for second click.
> 
> Your new version double click _**without my enhancement**_ active ![image](https://user-images.githubusercontent.com/12754913/241391027-1652a37a-92a2-418a-b757-159596e4b567.png)
> 
> Current pre-release double click **with a latex file** should go to this line in notepad ![image](https://user-images.githubusercontent.com/12754913/241391160-ec2008be-5046-4b85-94b1-002985120345.png)
> 
> I attach a demo set so you can see **normal pre-release behaviour** unpack this ZIP and drop sync.pdf onto the pre-release exe then double click a word should open notepad with the source line. (It will be **@start of non highlighted notepad line** as I have not decided what my latex editor should do.) In a real LaTeX IDE the behaviour can be variable.
> 
> [LATeX and Reverse-Search-textpad-demo.zip](https://github.com/sumatrapdfreader/sumatrapdf/files/11581926/LATeX.and.Reverse-Search-textpad-demo.zip)

![2023-05-27 220452](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/f992527f-3b4d-4436-9004-96e2df4394a5)

I have uploaded the executable file that solves the problem to my GitHub. The left mouse button click and double-click now functions as it did originally. The "Free text" can be selected from the context menu that appears when you right-click the mouse.  Thanks for your feedback.

https://github.com/youmjb/SumatraPDF-mod

**Reply by GitHubRulesOK** [reply]
OK that works well with Latex sample OR when latex is off my AHK select by drag also works,
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/d6c7c5c7-f278-41cb-a782-9b09fd719071)

Originally KJK has selection Auto populate the highlight but after comments that was dropped however some users need to drag  a selection and use that as the comment.

you can see most of those enhancements are now part of pre-release so only click and drag is worth keeping.

**Reply by youmjb** [reply]
Hi! I always regretted that there is no function to resize objects with mouse in sumatraPDF. I've been working on this for the past few days and here are the results. 

1. I added highlight, underline, squiggly, and textbox features to the menu bar.

![mpv-shot0009](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/32148f9b-7ab0-4717-bf60-6689e61d65b2)
![mpv-shot0010](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/f5e79c1e-7e8b-4700-be41-f3eddde37d88)
![mpv-shot0011](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/8156157a-16f3-4fb3-87e1-d13f3e57b57c)


2. Now you can resize and move any object including circle, square, caret, stamp, and image using the mouse. Even line and free text annotations are now resizable.

![mpv-shot0015](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/b415d038-5f07-43a3-920b-154eb098d3d6)
![mpv-shot0016](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/d1401947-824c-479b-a3d5-0cea9e922654)
![mpv-shot0017](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/ab969e55-afa5-4519-98ea-bbd0d0c01112)
![mpv-shot0018](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/9cfb5ab7-dbe3-4295-8f92-a1c14a023c44)
![mpv-shot0020](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/fd867647-0bf9-4563-bece-7ee29e239491)
![mpv-shot0024](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/9b04b27e-cad2-4a83-83fe-b52487f9de8f)
![mpv-shot0025](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/8024dc74-15e8-4e4c-b0db-e5aaa716d084)
![mpv-shot0026](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/30a01491-0690-49c8-a3f0-9914da87fc37)
![mpv-shot0028](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/414b05c4-686c-4734-982f-d1221acf720a)
![mpv-shot0029](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/c1690a6d-d605-43d3-9ec5-b13650212ae1)
![mpv-shot0014](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/74230f75-4021-4405-b2bb-67988521bf04)


3. You can now resize free text annotations using the mouse. I also removed line breaks when copying text from PDF documents to avoid unnecessary sentence breaks.  

![mpv-shot0054](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/a9e6df31-2562-43bc-a20d-0031c31249ab)
![mpv-shot0057](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/3e232114-d66d-4287-946c-ec51e7793ba2)
![mpv-shot0063](https://github.com/sumatrapdfreader/sumatrapdf/assets/57849476/d2e5ee1b-bc64-471f-8b6c-8c6f12d8f354)

https://github.com/youmjb/SumatraPDF-mod

Your feedback is always welcome and appreciated.

**Reply by xmfbraga** [reply]
Hi, @youmjb. I appreciated very much the modifications you introduced to Sumatra, especially, for the free text resizing functionality. Congrats! However, you may not have noticed, but the text option is always placing the icon at the top of the page and no longer at the cursor position, as it used to be.

**Reply by youmjb** [reply]
> Hi, @youmjb. I appreciated very much the modifications you introduced to Sumatra, especially, for the free text resizing functionality. Congrats! However, you may not have noticed, but the text option is always placing the icon at the top of the page and no longer at the cursor position, as it used to be.

Thank you so much for your feedback, I have identified and fixed the issue. You should be able to download it again from the link below.
https://github.com/youmjb/SumatraPDF-mod

SumatraPDF.exe : An executable file that starts free text annotion immediately when you double-click the left mouse button.
SumatraPDF(Raw).exe:   normal executable file

**Comment by GitHubRulesOK** [comment]
Personally cant add to previous comments since I am using currently 32bit Win 10 pre-release and yours is 64bit 
However that seems a big leap forwards towards clossing many issues

@kjk seems to be busy  on his many other projects 
so the fact that current SumatraPDF pre-release is fairly stable has good and bad aspects :-)

**Reply by youmjb** [reply]
> Personally cant add to previous comments since I am using currently 32bit Win 10 pre-release and yours is 64bit However that seems a big leap forwards towards clossing many issues
> 
> @kjk seems to be busy on his many other projects so the fact that current SumatraPDF pre-release is fairly stable has good and bad aspects :-)

Hi! I uploaded the 32bit version of SumatraPDF(SumatraPDF(32bit).exe) to my github.

https://github.com/youmjb/SumatraPDF-mod

**Reply by youmjb** [reply]
There was a significant bug (crashing) with free text annotation, so I have uploaded a revised version that fixes it.

https://github.com/youmjb/SumatraPDF-mod

**Reply by GitHubRulesOK** [reply]
personally thanks for 32 bit, 
have not had time to review but was stable enough to test it functioned for stamp, had not yet played with text much.

**Comment by GitHubRulesOK** [comment]
OK have a bit of time to test
Stamping seems to have some scalar issues and cant understand why

Take a 100% actual size clip of your PDF (simpler to compare same source)
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/f5552bf0-5c90-432d-8f16-059e031f44a5)


The clip when seen by Paint was a bit odd shape so I altered CANVAS (properties) NOT PIXEL SCALE to a convenient 2" x 1" since then easy to use point sizes not dpi.

when pasted it comes in at 200 units by 100 units (placed to the right) not itself a problem but larger than expected
The size in points should be 144 x 72 (and as pixels should be 192 x 96)
If I alter size to 144 (second lower copy) it is close in scale but quality is still poorer than expected

I think for quality it perhaps should import at 192x96 pixels which needs boundary rescale on placement to 75% (72/96) without disturbing pixel size

**Reply by youmjb** [reply]
I have improved the code based on your feedback. Now, when copying and pasting images, they are pasted in their original size. However, the image quality still seems to be poor, possibly due to using clipboard images. As an alternative, I suggest that copying and pasting images while zoomed in on the PDF document results in much better image quality.

https://github.com/youmjb/SumatraPDF-mod

**Reply by GitHubRulesOK** [reply]
Thanks for your perseverance and tenacity as your possibly finding its no easy task to manage PDF oddities.

I had clipped at 200% to attempt ensure quality but there are many factors due to defaults so clipping is related to the source relative screen image so deliberately suggested use clip @ 100% as a baseline for understand behaviours,

Windows default image should be at 96 dpi hence the estimated size of 2 inches wide by one inch tall should be 192 x 96 pixels and using MS paint confirms the 96 dpi (but in win 11 may state 120 dpi) or paste at 100% allows adjust boundary  in all such cases the pixels should I hope be 196 x 96 NOT 120/96    I.E. I hope clipping has not changed to 240 x 120  but then its MS prerogative.

Now when placing picture PDF does not use dpi (all pixels should be unitless) so in PDF terms `/width 196 /height 96` can be any scale or ratio (often such a fax tiff could be "square" as 2:1 draft)  and thus for placement in default 72 points per inch the placement needs be 144 x 72 but result should be perfect as "unaltered" at 1:1 scale. 

When compare by zoom in there there will be perceptual difference as an image is not vectors so the real text will get better but image will not.

So as noticed Clip @ 200% should be naturally better when pasted at same scale however key point is at a given scale both should look similar.

**Comment by GitHubRulesOK** [comment]
@kjk 

I have used these enhancements in 32bit without any major new issue  so It would be great if these enhancements were added to current pre-release

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/0eda1750-dc40-4e5d-8f83-d0d792090b41)

**Comment by vadimkantorov** [comment]
@youmjb Could you please submit your changes as a PR? Hopefully, this would start a review and could end in upstreaming your useful features

**Reply by youmjb** [reply]
Thank you for your response. I appreciate your suggestion for a pull request. Due to some urgent commitments, I may not be able to work on this until after March. I will plan to break down the changes into smaller PRs to make the review process more manageable.

In the meantime, I've added a new installation file that includes several progress updates which you may find useful for reference.

I'll keep you updated on my progress.

**Comment by GitHubRulesOK** [comment]
Just a comment that there are more than a handful of ways to place images on a PDF surface.
However I scripted one way to overlay secondary contents thus **NOT annotations or widgets** but an overlay like 1960/70's CAD Mylars (boy does that show my pre-PDF age)
The contents are appended to each given page or referenced in multiple pages with one image for a session. However, this variation allows multiple images in the session so the file WILL get bigger by the density of the image sources (PDF does not use DPI so it the image is 1000 x 1000 pixels, and you place as 72points x 72points it will be 1 million pixels per square inch.) If an image is transparent like the one on the right or a signature PNG it will be see-through. https://github.com/GitHubRulesOK/SumatraPDF-Plus#putimage-new

<img width="1342" height="609" alt="image" src="https://github.com/user-attachments/assets/56042824-8b7c-4cfe-95c0-14fbd345a2f7" />
